### PR TITLE
Wip/ji/dbif

### PR DESCRIPTION
**Pull request by jacek-izykowski**
It's possible now to open files and manipulate chunks from multiple clients.
No parsers working yet and there are still things to improve, but it may be a good moment to have a look (or even merge).

### CODE REVIEW

**Comment by adamiwaniuk** [comment]
crash on chunk creation:

 	Qt5Core.dll!0000000056b1bd8a()	Unknown
>	veles.exe!veles::dbif::InfoPromise::gotInfo(QSharedPointer<veles::dbif::InfoReply> _t1) Line 141	C++
 	veles.exe!veles::client::NCWrapper::handleGetListReplyMessage(std::shared_ptr<veles::messages::MsgpackMsg> message) Line 265	C++
 	veles.exe!veles::client::NCWrapper::messageReceived(std::shared_ptr<veles::messages::MsgpackMsg> message) Line 921	C++
 	veles.exe!QtPrivate::QSlotObject<void (__cdecl veles::client::NCWrapper::*)(std::shared_ptr<veles::messages::MsgpackMsg>) __ptr64,QtPrivate::List<std::shared_ptr<veles::messages::MsgpackMsg> >,void>::impl(int which, QtPrivate::QSlotObjectBase * this_, QObject * r, void * * a, bool * ret) Line 149	C++
 	Qt5Core.dll!0000000056b1c1f2()	Unknown
 	veles.exe!veles::client::NetworkClient::messageReceived(std::shared_ptr<veles::messages::MsgpackMsg> _t1) Line 194	C++
 	veles.exe!veles::client::NetworkClient::newDataAvailable() Line 444	C++
 	Qt5Core.dll!0000000056b1c1f2()	Unknown
 	Qt5Network.dll!0000000056f2a582()	Unknown
 	Qt5Network.dll!0000000056f31627()	Unknown
 	Qt5Widgets.dll!0000000056fd78f0()	Unknown
 	Qt5Widgets.dll!0000000056fd6822()	Unknown
 	Qt5Core.dll!0000000056afd309()	Unknown
 	Qt5Core.dll!0000000056b43637()	Unknown
 	[External Code]	
 	Qt5Core.dll!0000000056b42ea6()	Unknown
 	qwindows.dll!00007ffb7d5dd639()	Unknown
 	Qt5Core.dll!0000000056af9b03()	Unknown
 	Qt5Core.dll!0000000056afc107()	Unknown
 	veles.exe!main(int argc, char * * argv) Line 74	C++
 	[External Code]

File: src/data/nodeid.cc [hunk 1]
```diff
@@ -74,6 +74,18 @@ bool NodeID::operator!=(const NodeID &other) const {
   return !(*this == other);
 }
 
+bool NodeID::operator<(const NodeID& other) const {
+  for(int i = 0; i < WIDTH; ++i) {
```
Comment (adamiwaniuk) [inline src/data/nodeid.cc]: comparison between signed and unsigned integer expressions [-Wsign-compare]

**Comment by adamiwaniuk** [comment]
When using 200 mb file sometimes it works, sometimes it faield to upload:
Exception in callback _SelectorSocketTransport._read_ready()
handle: <Handle _SelectorSocketTransport._read_ready()>
Traceback (most recent call last):
  File "C:\Users\adam\AppData\Local\Programs\Python\Python36-32\lib\asyncio\events.py", line 127, in _run
    self._callback(*self._args)
  File "C:\Users\adam\AppData\Local\Programs\Python\Python36-32\lib\asyncio\selector_events.py", line 731, in _read_ready
    self._protocol.data_received(data)
  File "C:\Users\adam\dev\a\veles\python\veles\server\proto.py", line 261, in data_received
    msg = messages.MsgpackMsg.load(self.unpacker.unpack())
  File "msgpack/_unpacker.pyx", line 437, in msgpack._unpacker.Unpacker.unpack (msgpack/_unpacker.cpp:5037)
  File "msgpack/_unpacker.pyx", line 400, in msgpack._unpacker.Unpacker._unpack (msgpack/_unpacker.cpp:4483)
MemoryError



Sometimes it faild to load.
Sometimes it crashes:
 	Qt5Core.dll!0000000056b1bd8a()	Unknown
>	veles.exe!veles::dbif::InfoPromise::gotInfo(QSharedPointer<veles::dbif::InfoReply> _t1) Line 141	C++
 	veles.exe!veles::client::NCWrapper::handleGetBinDataReplyMessage(std::shared_ptr<veles::messages::MsgpackMsg> message) Line 454	C++
 	veles.exe!veles::client::NCWrapper::messageReceived(std::shared_ptr<veles::messages::MsgpackMsg> message) Line 921	C++
 	veles.exe!QtPrivate::QSlotObject<void (__cdecl veles::client::NCWrapper::*)(std::shared_ptr<veles::messages::MsgpackMsg>) __ptr64,QtPrivate::List<std::shared_ptr<veles::messages::MsgpackMsg> >,void>::impl(int which, QtPrivate::QSlotObjectBase * this_, QObject * r, void * * a, bool * ret) Line 149	C++
 	Qt5Core.dll!0000000056b1c1f2()	Unknown
 	veles.exe!veles::client::NetworkClient::messageReceived(std::shared_ptr<veles::messages::MsgpackMsg> _t1) Line 194	C++
 	veles.exe!veles::client::NetworkClient::newDataAvailable() Line 444	C++
 	Qt5Core.dll!0000000056b1c1f2()	Unknown
 	Qt5Network.dll!0000000056f2a582()	Unknown
 	Qt5Network.dll!0000000056f31627()	Unknown
 	Qt5Widgets.dll!0000000056fd78f0()	Unknown
 	Qt5Widgets.dll!0000000056fd6822()	Unknown
 	Qt5Core.dll!0000000056afd309()	Unknown
 	Qt5Core.dll!0000000056b43637()	Unknown
 	[External Code]	
 	Qt5Core.dll!0000000056b42ea6()	Unknown
 	qwindows.dll!00007ffb7d5dd639()	Unknown
 	Qt5Core.dll!0000000056af9b03()	Unknown
 	Qt5Core.dll!0000000056afc107()	Unknown
 	veles.exe!main(int argc, char * * argv) Line 74	C++
 	[External Code]

File: include/client/dbif.h [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+      root_children_promises_;
+  std::unordered_map<uint64_t, QSharedPointer<NCObjectHandle>> created_objs_waiting_for_ack_;
+  typedef std::map<data::NodeID, NCObjectHandle> ChildrenMap;
+  std::unordered_map<uint64_t, QSharedPointer<ChildrenMap>> children_maps_;
+  std::unordered_set<uint64_t> subchunk_data_requests_;
+
+  bool detailed_debug_info_;
+};
+
+} // client
```
Comment (altran01) [inline include/client/dbif.h]: Not proper namespace ending comment - @mkow will eat you alive :>.

File: src/data/nodeid.cc (hunk 1 above)
Comment (jacek-izykowski) [inline src/data/nodeid.cc]: Fixed.

**Comment by jacek-izykowski** [comment]
@adamiwaniuk Crash on chunk creation (seemingly) fixed.

File: src/client/dbif.cc [hunk 3]
```diff
@@ -0,0 +1,943 @@
...
+void NCWrapper::handleRequestAckMessage(
+    std::shared_ptr<messages::MsgpackMsg> message) {
+  auto reply = std::dynamic_pointer_cast<messages::MsgRequestAck>(message);
+  if (reply) {
+    if (nc_->output() && detailed_debug_info_) {
+      *nc_->output() << QString("NCWrapper: received MsgRequestAck.") << endl;
+    }
+
+    const auto promise_iter = method_promises_.find(reply->rid);
+    if (promise_iter != method_promises_.end()  && promise_iter->second) {
```
Comment (adamiwaniuk) [inline src/client/dbif.cc]: number of spaces around && are inconsistent

File: src/client/dbif.cc [hunk 4]
```diff
@@ -0,0 +1,943 @@
...
+#include "client/dbif.h"
+
+namespace veles {
+namespace client {
+
+/*****************************************************************************/
+/* NCObjectHandle */
+/*****************************************************************************/
+
+NCObjectHandle::NCObjectHandle(NCWrapper* nc, data::NodeID id,
```
Comment (adamiwaniuk) [inline src/client/dbif.cc]: since this takes ownership of nc, I think parent of it should be changed to this

File: include/client/dbif.h (hunk 2 above)
Comment (jacek-izykowski) [inline include/client/dbif.h]: Fixed.

File: src/client/dbif.cc [hunk 5]
```diff
@@ -0,0 +1,943 @@
...
+    //TODO
+  }
+}
+
+void NCWrapper::handleQueryErrorMessage(msg_ptr message) {
+  auto reply = std::dynamic_pointer_cast<messages::MsgQueryError>(message);
+  if (reply) {
+    if (nc_->output() && detailed_debug_info_) {
+      *nc_->output()
+          << QString("NCWrapper: received MsgQueryError.") << endl
```
Comment (jacek-izykowski) [inline src/client/dbif.cc]: I like explicit information about an actual type.
Changed anyway.

File: src/client/dbif.cc [hunk 6]
```diff
@@ -0,0 +1,943 @@
...
+void NCWrapper::handleGetReplyMessage(
+    std::shared_ptr<messages::MsgpackMsg> message) {
+  auto reply = std::dynamic_pointer_cast<messages::MsgGetReply>(message);
+  if (reply) {
+    const auto promise_iter = promises_.find(reply->qid);
+    if (promise_iter != promises_.end() && promise_iter->second) {
+      QString name("");
+      QString comment("");
+
+      getQStringAttr(reply->obj->attr, std::string("name"), name);
```
Comment (jacek-izykowski) [inline src/client/dbif.cc]: I like explicit information about an actual type.
Changed anyway.

File: src/client/dbif.cc [hunk 7]
```diff
@@ -0,0 +1,943 @@
...
+
+template<class T> void getAttrSPtr(std::shared_ptr<std::unordered_map<
+    std::string, std::shared_ptr<messages::MsgpackObject>>> attr,
+    std::string key, T& val) {
+  auto iter = attr->find(key);
+  if (iter != attr->end()) {
+    std::shared_ptr<T> val_ptr;
+    messages::fromMsgpackObject(iter->second, val_ptr);
+    if (val_ptr) {
+      val = *val_ptr;
```
Comment (jacek-izykowski) [inline src/client/dbif.cc]: We generally don't send pointers through network. We expect messages to contain simple values or something that is easy to handle/copy.

File: src/client/dbif.cc [hunk 8]
```diff
@@ -0,0 +1,943 @@
...
+  }
+}
+
+void getQStringAttr(std::shared_ptr<std::unordered_map<std::string,
+    std::shared_ptr<messages::MsgpackObject>>> attr, std::string key,
+    QString& val) {
+  auto iter = attr->find(key);
+  if (iter != attr->end()) {
+    std::shared_ptr<std::string> val_ptr;
+    messages::fromMsgpackObject(iter->second, val_ptr);
```
Comment (jacek-izykowski) [inline src/client/dbif.cc]: Changed.

File: src/client/dbif.cc [hunk 9]
```diff
@@ -0,0 +1,943 @@
...
+    }
+
+    // TODO Do we have other cases when we need to do something special when
+    // MsgRequestAck is received?
+  }
+}
+
+void getQStringAttr(std::shared_ptr<std::unordered_map<std::string,
+    std::shared_ptr<messages::MsgpackObject>>> attr, std::string key,
+    QString& val) {
```
Comment (jacek-izykowski) [inline src/client/dbif.cc]: Changed. val -> val_out.

File: src/client/dbif.cc [hunk 10]
```diff
@@ -0,0 +1,943 @@
...
+}
+
+/*****************************************************************************/
+/* NCWrapper - NetworkClient message handlers */
+/*****************************************************************************/
+
+void NCWrapper::handleGetListReplyMessage(
+    std::shared_ptr<messages::MsgpackMsg> message) {
+  auto reply = std::dynamic_pointer_cast<messages::MsgGetListReply>(message);
+  if (reply) {
```
Comment (jacek-izykowski) [inline src/client/dbif.cc]: Fixed.

File: src/client/dbif.cc [hunk 11]
```diff
@@ -802,8 +805,8 @@ dbif::MethodResultPromise* NCWrapper::handleSetCommentRequest(data::NodeID id,
 
   if(nc_->connectionStatus() == NetworkClient::ConnectionStatus::Connected) {
     if (nc_->output() && detailed_debug_info_) {
-      *nc_->output() << QString("NCWrapper: Sending a request to set a node's"
-          "comment (MsgTransaction).") << endl;
+      *nc_->output() << "NCWrapper: Sending a request to set a node's"
+          "comment (MsgTransaction)." << endl;
```
Comment (adamiwaniuk) [inline src/client/dbif.cc]: please add space before "comment" here and "bounds" in line 845

**Comment by adamiwaniuk** [comment]
non ASCII characters are not displayed correctly in log window
logs

**Comment by altran01** [comment]
I think that when client is not connected to any server we probably shouldn't show database view just yet, and disallow open action which silently fails if we aren't connected anywhere.

File: src/client/dbif.cc [hunk 12]
```diff
@@ -0,0 +1,953 @@
...
+
+    connect(nc_, &NetworkClient::connectionStatusChanged,
+        this, &NCWrapper::updateConnectionStatus);
+    connect(nc_, &NetworkClient::messageReceived,
+        this, &NCWrapper::messageReceived);
+  }
+}
+
+dbif::ObjectType NCWrapper::typeFromTags(
+      std::shared_ptr<std::unordered_set<std::shared_ptr<std::string>>> tags) {
```
Comment (altran01) [inline src/client/dbif.cc]: Eh, now that I see it used, I see how it sucks. We either need to remove pointers from strings or provide more template parameters to this template to make it usable in any way - I'll prepare a patch with removing pointers for now.

**Comment by altran01** [comment]
BTW, can you make make_blob and upload_elf add files in a way that allows for them to be displayed in the UI?

**Comment by jacek-izykowski** [comment]
@altran01 make_blob and upload_elf are updated in commit 5202ffabd096dd973e8b2c5dc2969083914c6c6f.

**Comment by jacek-izykowski** [comment]
@altran01 Database tab and "Open" action are now disabled when not connected (commit 1a766ab3a1e681df9dea7647d25ffdb4539c6671).

**Comment by jacek-izykowski** [comment]
@adamiwaniuk Encoding issue fixed in 078db2ab475629c4ed22dfde19aa7f3e99e31f65.

**Pull request merged**